### DISCUSSION TITLE
PDF Background

### DISCUSSION THREAD

**Discussion by pplotka** [gotenberg/gotenberg#461] [Q&A]
Hi! 

I would like to ask you about the pdf background feature. I have to generate fulfilled forms. I have templates of these forms in pdf format. I used HTML (wkhtmltopdf and pdftk separately to generate fulfilled form). wkhtmltopdf generate pdf and then pdftk set background from template pdf file.
Now we rewrite almost the whole pdf generation to use Gotenberg and it would be great to have the possibility to set the background from the PDF file. I try to understand how to extend/customize Gotenberg but I don't know. Maybe you have any examples or tips on how to achieve described feature.

**Comment by gulien** [comment]
Hello @pplotka 👋 

Would it be possible to set the background using HTML + CSS? 😄

**Reply by pplotka** [reply]
Unfortunately, this isn't easy. I have a problem with scaling and positioning text on this form template. Another consequence of this is the poor quality of generated PDFs.

**Reply by gulien** [reply]
@pplotka I currently do not have time to implement such feature, but you could open a PR with:

* a new method interface (for instance, `AddBackground` or `WithBackground`) for the [PDFEngine](https://github.com/gotenberg/gotenberg/blob/main/pkg/gotenberg/pdfengine.go#L34) interface.
* implement it in the [PDFtk](https://github.com/gotenberg/gotenberg/blob/main/pkg/modules/pdftk/pdftk.go) module, but also in the [PDF engines wrapper](https://github.com/gotenberg/gotenberg/blob/main/pkg/modules/pdfengines/multi.go). Note: for other PDF engine implementation, you also have to implement it by returning [a specific error](https://github.com/gotenberg/gotenberg/blob/main/pkg/modules/pdfcpu/pdfcpu.go#L54).

From here, you may add:

* a new route to the [PDFEngines module](https://github.com/gotenberg/gotenberg/blob/main/pkg/modules/pdfengines/pdfengines.go#L166).
* add it as an optional form field for the `merge` and `convert` routes (https://github.com/gotenberg/gotenberg/blob/main/pkg/modules/pdfengines/routes.go).
